### ISSUE TITLE
Different results of eval() and forward().

### ISSUE THREAD

**Issue by ivan-alles** [Microsoft/CNTK#3229]
I created a model to gradients wrt. the input variable. To do that, I computed forward() and backward(), as described in CNTK 205: Artistic Style Transfer tutorial. At some point I noticed that forward() returns a wrong result (different from eval()). 

I reproduces this behavior using a CNTK example TransferLearning.py. I will add the full adapted file into an attachment. Now I just post the changed lines:

    # compute model output
    arguments = {loaded_model.arguments[0]: [hwc_format]}
    output = loaded_model.eval(arguments)

    # return softmax probabilities
    sm = softmax(output[0])

    #-------------- Begin changed ----------------------

    # Clone a model allowing gradient computation
    old_input_var = loaded_model.signature[0]

    cloned = loaded_model.clone(C.CloneMethod.freeze, {old_input_var : C.placeholder() })

    new_input_var = C.ops.input_variable(old_input_var.shape, np.float32, name='input_var', needs_gradient=True)
    new_model = cloned(new_input_var)

    new_output = new_model.eval(hwc_format)

    #keep_for_backward = None # This generates the same output
    keep_for_backward = {new_model.output} # This one generates a different output

    backprop_state, forward_output = new_model.forward(hwc_format, keep_for_backward=keep_for_backward)
    forward_output = forward_output[new_model.output]

    print('output        ', output[0])
    print('new output    ', new_output[0])
    print('forward_output', forward_output[0]) # This one differs !!!!!!!!!!!!!

    return sm.eval()
    #-------------- End changed  ----------------------

The original model's and the new model's eval() are the same. The new models' forward() is different (as soon as I specify something in keep_for_forward).

In my opinion, this is because the model (ResNet18) uses BatchNormalization layers, which behaves differently in eval() and forward(). 

I assume that this will also lead to wrong gradient computation in backward(). 

Are my assumptions correct? How can I fix this? 

I could replace BatchNormalization to fixed-weight layers, but I do not know where to get the long-term estimations for the mean/std.

**Comment by bencherian** [comment]
This is expected behavior. CNTK uses the `keep_for_backward` to determine if the model is running in "training" or "testing" mode. (If there are any `keep_for_backward` elements, the model is assumed to be in training.) BatchNorm uses different parameter values for mean/sigma calculations depending on whether or not you are training or testing, which explains the different behavior in your code.

**Comment by ivan-alles** [comment]
Thank you for the explanation. So my initial guess was correct. But I cannot agree that this is an expected behavior. If I clone a model with `CloneMethod.freeze`, I expect that the weights will not change, regardless of the mode (training or evaluation).

It is important to freeze the weights for both transfer learning and all algorithms requiring differentiation wrt. to inputs (like Artistic Style Transfer). It looks like the corresponding examples and tutorials on CNTK does not work as expected and change the base model during execution. 

I suggests that you add a way to fix the weights as a long-term solution.

As a workaround I could replace `BatchNormalization` layers with some linear transformation. I assume that mean and sigma are related to `scale` and `bias`parameters of the BN. Can you tell me the exact formula which is used in the BN?

**Comment by KeDengMS** [comment]
Forward in training is different than eval in batch normalization, as training updates statistics while eval does not. For BN implementation details, please take a look at [this test](https://github.com/Microsoft/CNTK/blob/master/bindings/python/cntk/ops/tests/non_linear_test.py#L616-L679).
For BN in CPU, which version of CNTK are you using? BN training on CPU has been in for a while.
For DropOut, you can call `drop_out_node.set_attribute('dropOutRate', 0.0)` to change the rate.

If you'd like to eval on some graph, while train on another, you can do eval(..., as_numpy=False) to get a GPU value object that can be feed into the training network.

**Comment by ivan-alles** [comment]
Thank you for the suggestions, now I see how I can do transfer learning and deactivate the dropout. 
I'm working with CNTK 2.4, CPU training (actually forward/backward) does not work for BN.

**Comment by KeDengMS** [comment]
2.4 should have BN training on CPU with [this commit](https://github.com/Microsoft/CNTK/commit/1d3d1e8a2770b65a809f6bf7e6cde49047ffc0c6) using MKL. Are you not using MKL?

**Comment by ivan-alles** [comment]
I double-checked it. For CPU we used the CNTK python distribution 2.1. Now we switched to the python distribution 2.4 and it works.
